src/day8/re_frame/test.cljc

```diff
@@ -1,66 +1,242 @@
+;;;; General test utils
+;;;;
+
+(defn- dequeue!
+  "Dequeue an item from a persistent queue which is stored as the value of the
+  given atom, setting the atom to the new queue (minus dequeued item) and
+  returning the dequeued item. If the queue is empty, does not alter it and
+  returns nil."
+  [queue-atom]
+  (let [queue @queue-atom]
+    (when (seq queue)
+      (if (compare-and-set! queue-atom queue (pop queue))
+        (peek queue)
+        (recur queue-atom)))))
+
+
+(defn make-re-frame-restore-fn []
+  (let [handlers @@#'rf-registrar/kind->id->handler
+        app-db   @@#'rf-db/app-db]
+    (fn []
+      (reset! @#'rf-registrar/kind->id->handler handlers)
+      (reset! @#'rf-db/app-db app-db)
+      nil)))
+
+(defmacro with-restored-re-frame-state
+  "Run `body`, but discard whatever effects it may have on re-frame's internal
+  state (by resetting `app-db` and re-frame's various different types of
+  handlers after `body` has run)."
+  [& body]
+  `(let [restore-fn# (make-re-frame-restore-fn)]
+     (try
+       ~@body
+       (finally
+         (restore-fn#)))))
+
+
+;;;;
+;;;; Async tests
+;;;;
+
+(def ^:dynamic *test-timeout* 5000)
+
+(def ^{:dynamic true, :private true} *test-context*
+  "`*test-context*` is used to communicate internal details of the test between
+  `run-test-async*` and `wait-for*`. It is dynamically bound so that it doesn't
+  need to appear as a lexical argument to a `wait-for` block, since we don't
+  want it to be visible when you're writing tests.  But care must be taken to
+  pass it around lexically across callbacks, since ClojureScript doesn't have
+  `bound-fn`."
+  nil)
+
+
+(defn run-test-async* [f]
+  (let [test-context {:wait-for-depth     0
+                      :max-wait-for-depth (atom 0)
+                      :now-waiting-for    (atom nil)}]
+    #?(:clj  (let [done-promise (promise)
+                   executor     (Executors/newSingleThreadExecutor)
+                   fail-ex      (atom nil)
+                   test-context (assoc test-context :done #(deliver done-promise ::done))]
+               (with-redefs [rf-int/executor executor]
+                 ;; Execute the test code itself on the same thread as the
+                 ;; re-frame event handlers run, so that we accurately
+                 ;; simulate the single-threaded JS environment and also so
+                 ;; that we don't have to worry about making the JVM
+                 ;; implementation of the re-frame EventQueue thread-safe.
+                 (rf-int/next-tick #(try
+                                      (binding [*test-context* test-context]
+                                        (f))
+                                      (catch Throwable t
+                                        (reset! fail-ex t)))))
+               (let [result (deref done-promise *test-timeout* ::timeout)]
+                 (doto executor
+                   (.shutdownNow)
+                   (.awaitTermination 1 TimeUnit/SECONDS))
```

Comment:
I struggle to see it happening in practice (you'd have to have an event handler which caught `InterruptedException` and still didn't return), but if it does, I suppose that would mean that you could have a process which is ultimately going to `reset!` the `app-db` *after* the `with-temp-re-frame-state` macro has cleaned up after itself.  So might be better to be told of that via an exception from an otherwise-passing test.